Plugins/Public/mobiledocking_plugin/Main.cpp

```diff
@@ -15,21 +15,57 @@ vector<uint> dockingModuleEquipmentIds;
 // Above how much cargo capacity, should a ship be rejected as a docking user?
 int cargoCapacityLimit = 275;
 
+// How much time will player be given before kick if carrier wants to jettison him.
+int jettisonKickTime = 15;
+
 uint connSystemID = CreateID("li06");
 
-///////////////////////////////////////////////////////////////////////////////////////////////////////////////
-// Function handling changing systems when undocking from a carrier
-void JumpToLocation(uint client, uint system, Vector pos, Matrix ornt)
+// Delayed actions, which need to be done. Look at HkTimerCheckKick().
+vector<ActionJettison> jettisonList;
+
+void HkTimerCheckKick()
+{
+	returncode = DEFAULT_RETURNCODE;
+
+	for (vector<ActionJettison>::iterator it = jettisonList.begin(); it != jettisonList.end(); it++)
+	{
+		it->timeLeft--;
+		if (it->timeLeft == 0)
+		{
+			uint checkCarrierClientID = HkGetClientIdFromCharname(it->carrierCharname);
+			uint checkDockedClientID = HkGetClientIdFromCharname(it->dockedCharname);
+
+			// If both carrier and docked ship are still at server.
+			if (checkDockedClientID != -1 && checkCarrierClientID != -1)
+			{
+				if (mobiledockClients[checkDockedClientID].wscDockedWithCharname == it->carrierCharname)
+				{
+					mobiledockClients[checkCarrierClientID].mapDockedShips.erase(it->dockedCharname);
+					HkKick(HkGetAccountByClientID(checkDockedClientID));
+					PrintUserCmdText(checkCarrierClientID, L"Ship jettisoned.");
+				}
+			}
+
+			jettisonList.erase(it);
+		}
+	}
+}
+
+void JettisonShip(uint carrierClientID, uint dockedClientID)
 {
-	mapDeferredJumps[client].system = system;
-	mapDeferredJumps[client].pos = pos;
-	mapDeferredJumps[client].rot = ornt;
+	const wchar_t* carrierCharname = (const wchar_t*)Players.GetActiveCharacterName(carrierClientID);
+	const wchar_t* dockedCharname = (const wchar_t*)Players.GetActiveCharacterName(dockedClientID);
 
-	// Send the jump command to the client. The client will send a system switch out complete
-	// event which we intercept to set the new starting positions.
-	PrintUserCmdText(client, L" ChangeSys %u", system);
+	if (dockedClientID != -1)
+	{
+		PrintUserCmdText(carrierClientID, L"Ship warned. If it won't undock in %i seconds, it will be kicked by force.", jettisonKickTime);
+		PrintUserCmdText(dockedClientID, L"Carrier wants to jettison your ship. Undock willingly or you will be kicked after %i seconds.", jettisonKickTime);

```

Comment:
or carrier has died?

Reply:
Same messages in cases if carrier died or wants to jettison ship - is not a problem.

Reply:
No, this is a problem. When the carrier dies the message docked clients are told will be a lie. The message should read, based on the current callers, 'Either the carrier wants to jettison your ship, or it has died.'

Reply:
Solved.